cpp/src/pylogging.cpp

```diff
@@ -35,12 +35,30 @@ py::object CppToPythonLogger::getLogger() {
     return logger_;
 }
 
+/// Saves error and restores it at the end of the scope,
+/// unless another one has been set in the meantime.
+struct save_python_error {
+    PyObject *type, *value, *trace;
+    save_python_error() { PyErr_Fetch(&type, &value, &trace); }
+
+    ~save_python_error() {
+        if (PyErr_Occurred() == nullptr) {
+            PyErr_Restore(type, value, trace); 
+        }
+    }
+};
+
 void logFromJava(int level, long timestamp, char* loggerName, char* message) {
+    py::gil_scoped_acquire acquire;
+    save_python_error previousError;  // to keep and restore the previously set exception, if any
```

Comment:
@obrix I added this mechanism to allow to keep the initial exception during the subsequent log calls.

It seems that without this, after the log call which sets the error, the previously set error is cleared (the python error is set at the beginning of the call, but not set anymore at the end ...).

With that additional workaround, the KeyboardInterrupt exception is correctly passed all the way until the end of the java call, and re-thrown thanks to your code.
From python side, we correctly receive it:
```
Traceback (most recent call last):
  File "toto.py", line 8, in <module>
    network = pn.load('/home/leclercsyl/feature_tests/01-test-cases/02-full/pf_20200326T1635Z_20200326T1635Z.xiidm')
  File "/home/leclercsyl/Projects/sources/powsybl/pypowsybl/pypowsybl/network.py", line 4058, in load
    return Network(_pp.load_network(file, parameters))
  File "/usr/lib/python3.8/logging/__init__.py", line 1497, in log
    def log(self, level, msg, *args, **kwargs):
KeyboardInterrupt
``` 

It looks good this way :)

Reply:
Cool, thank you!